spec/fixtures/custom_nginx.template

```diff
@@ -312,60 +312,156 @@ http {
+                    }
+                })
+            }
+        }
 
-                while true do
-                  local data, typ, err = wb:recv_frame()
-                  if wb.fatal then
-                    ngx.log(ngx.ERR, "failed to receive frame: ", err)
-                    return ngx.exit(444)
-                  end
+        location = /ws {
+            content_by_lua_block {
+                local mu = require('spec.fixtures.mock_upstream')
+                return mu.serve_web_sockets()
+            }
+        }
 
-                  if data then
-                    if typ == "close" then
-                      break
-                    end
-
-                    if typ == "ping" then
-                      local bytes, err = wb:send_pong(data)
-                      if not bytes then
-                        ngx.log(ngx.ERR, "failed to send pong: ", err)
-                        return ngx.exit(444)
-                      end
-
-                    elseif typ == "pong" then
-                      ngx.log(ngx.INFO, "client ponged")
-
-                    elseif typ == "text" then
-                      local bytes, err = wb:send_text(data)
-                      if not bytes then
-                        ngx.log(ngx.ERR, "failed to send text: ", err)
-                        return ngx.exit(444)
-                      end
-                    end
+        location = /get {
+            access_by_lua_block {
+                local mu = require('spec.fixtures.mock_upstream')
+                return mu.filter_access_by_method("GET")
+            }
+            content_by_lua_block {
+                local mu = require('spec.fixtures.mock_upstream')
+
+                return mu.send_json_response({
+                    args    = ngx.req.get_uri_args(),
+                    headers = ngx.req.get_headers(),
+                    origin  = ngx.var.remote_addr,
+                    url     = mu.get_request_url(),
+                })
+            }
+        }
+        location = /post {
+            access_by_lua_block {
+                local mu = require('spec.fixtures.mock_upstream')
+                return mu.filter_access_by_method("POST")
+            }
+            content_by_lua_block {
+                local mu                 = require('spec.fixtures.mock_upstream')
+                local headers            = ngx.req.get_headers()
+                local data, form, params = mu.parse_post_data(headers)
+                return mu.send_json_response({
+                    args    = ngx.req.get_uri_args(),
+                    data    = data,
+                    form    = form,
+                    headers = headers,
+                    origin  = ngx.var.remote_addr,
+                    params  = params,
+                    url     = mu.get_request_url(),
+                })
+            }
+        }
+        location ~ "^/basic-auth$" {
+            return 301 /basic-auth/user/passwd;
```

Comment:
This was me being cute (httpbin does the redirect, but we don't need it for the tests). I have removed both redirects. It's as you say, less code, less errors. Thanks